// @google/adk/dist/cjs/utils/simple_zod_to_json.js
/**
  * @license
  * Copyright 2025 Google LLC
  * SPDX-License-Identifier: Apache-2.0
  */

"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var simple_zod_to_json_exports = {};
__export(simple_zod_to_json_exports, {
  isZodObject: () => isZodObject,
  zodObjectToSchema: () => zodObjectToSchema
});
module.exports = __toCommonJS(simple_zod_to_json_exports);
var import_genai = require("@google/genai");
var import_zod_to_json_schema = require("zod-to-json-schema");
var import_v4 = require("zod/v4");
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
function isZodSchema(obj) {
  return obj !== null && typeof obj === "object" && "parse" in obj && typeof obj.parse === "function" && "safeParse" in obj && typeof obj.safeParse === "function";
}
function isZodV3Schema(obj) {
  return isZodSchema(obj) && !("_zod" in obj);
}
function isZodV4Schema(obj) {
  return isZodSchema(obj) && "_zod" in obj;
}
function getZodTypeName(schema) {
  var _a, _b;
  const schemaAny = schema;
  if ((_a = schemaAny._def) == null ? void 0 : _a.typeName) {
    return schemaAny._def.typeName;
  }
  const zod4Type = (_b = schemaAny._def) == null ? void 0 : _b.type;
  if (typeof zod4Type === "string" && zod4Type) {
    return "Zod" + zod4Type.charAt(0).toUpperCase() + zod4Type.slice(1);
  }
  return void 0;
}
function isZodObject(obj) {
  return isZodSchema(obj) && getZodTypeName(obj) === "ZodObject";
}
function zodObjectToSchema(schema) {
  if (!isZodObject(schema)) {
    throw new Error("Expected a Zod Object");
  }
  if (isZodV4Schema(schema)) {
    return (0, import_v4.toJSONSchema)(schema, {
      target: "openapi-3.0",
      io: "input",
      override: (ctx) => {
        var _a;
        const { jsonSchema } = ctx;
        if (jsonSchema.additionalProperties !== void 0) {
          delete jsonSchema.additionalProperties;
        }
        if (jsonSchema.readOnly !== void 0) {
          delete jsonSchema.readOnly;
        }
        if (jsonSchema.maxItems !== void 0) {
          jsonSchema.maxItems = jsonSchema.maxItems.toString();
        }
        if (jsonSchema.format === "email" || jsonSchema.format === "uuid") {
          delete jsonSchema.pattern;
        }
        if (jsonSchema.minItems !== void 0) {
          jsonSchema.minItems = jsonSchema.minItems.toString();
        }
        if (jsonSchema.minLength !== void 0) {
          jsonSchema.minLength = jsonSchema.minLength.toString();
        }
        if (jsonSchema.maxLength !== void 0) {
          jsonSchema.maxLength = jsonSchema.maxLength.toString();
        }
        if (((_a = jsonSchema.enum) == null ? void 0 : _a.length) === 1 && jsonSchema.enum[0] === null) {
          jsonSchema.type = import_genai.Type.NULL;
          delete jsonSchema.enum;
        }
        if (jsonSchema.type !== void 0) {
          jsonSchema.type = jsonSchema.type.toUpperCase();
        }
      }
    });
  }
  if (isZodV3Schema(schema)) {
    return (0, import_zod_to_json_schema.zodToJsonSchema)(schema, {
      target: "openApi3",
      emailStrategy: "format:email",
      postProcess: (jsonSchema) => {
        var _a, _b, _c, _d, _e, _f, _g;
        if (!jsonSchema) {
          return;
        }
        if (jsonSchema.additionalProperties !== void 0) {
          delete jsonSchema.additionalProperties;
        }
        if (jsonSchema.maxItems !== void 0) {
          jsonSchema.maxItems = (_a = jsonSchema.maxItems) == null ? void 0 : _a.toString();
        }
        if (jsonSchema.minItems !== void 0) {
          jsonSchema.minItems = (_b = jsonSchema.minItems) == null ? void 0 : _b.toString();
        }
        if (jsonSchema.minLength !== void 0) {
          jsonSchema.minLength = (_c = jsonSchema.minLength) == null ? void 0 : _c.toString();
        }
        if (jsonSchema.maxLength !== void 0) {
          jsonSchema.maxLength = (_d = jsonSchema.maxLength) == null ? void 0 : _d.toString();
        }
        if (((_e = jsonSchema.enum) == null ? void 0 : _e.length) === 1 && jsonSchema.enum[0] === "null") {
          jsonSchema.type = import_genai.Type.NULL;
          delete jsonSchema.enum;
        }
        if (jsonSchema.type === "integer" && jsonSchema.format !== "int64") {
          (_f = jsonSchema.minimum) != null ? _f : jsonSchema.minimum = Number.MIN_SAFE_INTEGER;
          (_g = jsonSchema.maximum) != null ? _g : jsonSchema.maximum = Number.MAX_SAFE_INTEGER;
        }
        if (jsonSchema.type !== void 0) {
          jsonSchema.type = jsonSchema.type.toUpperCase();
        }
        return jsonSchema;
      }
    });
  }
  throw new Error("Unsupported Zod schema version.");
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  isZodObject,
  zodObjectToSchema
});
